rust/parser/src/number_state.rs

```diff
@@ -0,0 +1,231 @@
+    cards::ParserCard,
+    decoding_commons::OutputCard,
+    state::DefaultState,
+    state::{State, StateError, StateOutput},
+};
+
+use merkleized_metadata::ExtraInfo;
+
+use printing_balance::convert_balance_pretty;
+
+use num_bigint::{BigInt, BigUint};
+
+pub trait NumberCardProducing: Sync + Send + Clone {
+    fn produce_card(&self, number: String, indent: u32) -> OutputCard;
+}
+
+/// A decoding state that detects numeric values and outputs a single
+/// card, as defined by the provided `card_producer` implementation.
+///
+/// This approach allows for flexible customization of how numeric values
+/// are displayed based on their context. For example, a "tip" value shown
+/// in an extension may be rendered with a different style than a balance
+/// value inside a transfer call.
+///
+/// By abstracting card creation through `card_producer`, this state enables
+/// context-aware formatting and styling of numbers during decoding.
+#[derive(Debug, Clone)]
+pub struct NumberState<T: NumberCardProducing> {
+    pub card_producer: T,
+}
+
+impl<T: NumberCardProducing> NumberState<T> {
+    fn process_number(&self, number: String, indent: u32) -> StateOutput {
+        let card = self.card_producer.produce_card(number, indent);
+        StateOutput {
+            next_state: Box::new(DefaultState),
+            cards: vec![card],
+            indent,
+        }
+    }
+}
+
+impl<T: NumberCardProducing + 'static> State for NumberState<T> {
+    fn clone_box(&self) -> Box<dyn State> {
+        Box::new(self.clone())
+    }
+
+    fn process_u8(&self, input: u8, indent: u32) -> Result<StateOutput, StateError> {
+        let output = self.process_number(input.to_string(), indent);
+
+        Ok(output)
+    }
+
+    fn process_u16(&self, input: u16, indent: u32) -> Result<StateOutput, StateError> {
+        let output = self.process_number(input.to_string(), indent);
+
+        Ok(output)
+    }
+
+    fn process_u32(&self, input: u32, indent: u32) -> Result<StateOutput, StateError> {
+        let output = self.process_number(input.to_string(), indent);
+
+        Ok(output)
+    }
+
+    fn process_u64(&self, input: u64, indent: u32) -> Result<StateOutput, StateError> {
+        let output = self.process_number(input.to_string(), indent);
+
+        Ok(output)
+    }
+
+    fn process_u128(&self, input: u128, indent: u32) -> Result<StateOutput, StateError> {
+        let output = self.process_number(input.to_string(), indent);
+
+        Ok(output)
+    }
+
+    fn process_u256(&self, input: &[u8; 32], indent: u32) -> Result<StateOutput, StateError> {
+        let target_value = BigUint::from_bytes_le(input);
+        let output = self.process_number(target_value.to_string(), indent);
+
+        Ok(output)
+    }
+
+    fn process_i8(&self, input: i8, indent: u32) -> Result<StateOutput, StateError> {
+        let output = self.process_number(input.to_string(), indent);
+
+        Ok(output)
+    }
+
+    fn process_i16<'scale, 'resolver>(
```

Comment:
Do we need lifetimes here? I don't see them in other number methods, e.g. `i32` or `i8`

Reply:
I don't think so we need those parameters. Removed.